### PR TITLE
Added password support for SSL private key of the cached schema registry client

### PR DESCRIPTION
**Pull request by mahajanadhitya**

### CODE REVIEW

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 1]
```diff
@@ -159,10 +173,17 @@ def _send_request(self, url, method='GET', body=None, headers={}):
...
-            _headers["Content-Length"] = str(len(body))
-            _headers["Content-Type"] = "application/vnd.schemaregistry.v1+json"
+            nbody = json.dumps(body).encode('UTF-8')
+            _headers["Content-Length"] = str(len(nbody))
         _headers.update(headers)
+        if url.startswith('https'):
+            response = self._https_session.request(method, url, headers=_headers, body=nbody)
+            try:
+                return json.loads(response.data), response.status
+            except ValueError:
```
Comment (pranavrth) [inline src/confluent_kafka/avro/cached_schema_registry_client.py:185]: Test this part as well.

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 2]
```diff
@@ -100,14 +104,22 @@ def __init__(self, url, max_schemas_per_subject=1000, ca_location=None, cert_loc
         # subj => { schema => version }
         self.subject_to_schema_versions = defaultdict(dict)
 
+        ca_certs = conf.pop('ssl.ca.location', None)
```
Comment (pranavrth) [inline src/confluent_kafka/avro/cached_schema_registry_client.py:107]: Can this be empty? What is the behaviour in that case? Ideally it should refer to the default paths.

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 3]
```diff
@@ -159,10 +173,17 @@ def _send_request(self, url, method='GET', body=None, headers={}):
             raise ClientError("Method {} is invalid; valid methods include {}".format(method, VALID_METHODS))
 
         _headers = {'Accept': ACCEPT_HDR}
+        nbody = body
         if body:
-            _headers["Content-Length"] = str(len(body))
-            _headers["Content-Type"] = "application/vnd.schemaregistry.v1+json"
```
Comment (pranavrth) [inline src/confluent_kafka/avro/cached_schema_registry_client.py:164]: Is content type header not required anymore?

File: tests/integration/integration_test.py [hunk 4]
```diff
@@ -1150,7 +1150,8 @@ def resolve_envs(_conf):
             _conf[k] = os.getenv(v[1:])
 
 
-test_modes = ['consumer', 'producer', 'avro', 'performance', 'admin', 'avro-https', 'avro-basic-auth', 'throttle']
+test_modes = ['consumer', 'producer', 'avro', 'performance', 'admin',
```
Comment (pranavrth) [inline tests/integration/integration_test.py:1153]: Not needed

File: tests/docker/.env.sh [hunk 5]
```diff
@@ -1,15 +1,17 @@
 #!/usr/bin/env bash
 
-export PY_DOCKER_SOURCE="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null && pwd )"
+export PY_DOCKER_SOURCE=$(pwd)
```
Comment (pranavrth) [inline tests/docker/.env.sh:3]: @emasab would be better to comment on this and next file.

**Review by pranavrth — CHANGES_REQUESTED** [review]
Please do testing with missing parameters as well.

@emasab - Can you also have a look on this PR?

File: CHANGELOG.md [hunk 6]
```diff
@@ -6,6 +6,8 @@
   SASL PLAIN/SCRAM credentials that will be used for subsequent (new) connections to a broker (#1511).
 - Wheels for Linux / arm64 (#1496).
 - Added support for Default num_partitions in CreateTopics Admin API.
+- Added support for password of Private Key for SSL in CachedSchemaRegistryClient
```
[the remark this answers is not in the document]
Comment (mahajanadhitya) [reply CHANGELOG.md:9]: Sure.

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 7]
```diff
@@ -159,10 +173,17 @@ def _send_request(self, url, method='GET', body=None, headers={}):
...
+            nbody = json.dumps(body).encode('UTF-8')
+            _headers["Content-Length"] = str(len(nbody))
         _headers.update(headers)
+        if url.startswith('https'):
```
[the remark this answers is not in the document]
Comment (mahajanadhitya) [reply src/confluent_kafka/avro/cached_schema_registry_client.py:181]: yes, basically whenever the request is to https client it might have password protected key or not , either way this attribute works while the other only works for https SSL connection with non password protected key.

File: src/confluent_kafka/avro/cached_schema_registry_client.py (hunk 3 above)
Comment (mahajanadhitya) [reply src/confluent_kafka/avro/cached_schema_registry_client.py:164]: the constant have this already declared and the constant was declared before also

File: tests/integration/integration_test.py (hunk 4 above)
Comment (mahajanadhitya) [reply tests/integration/integration_test.py:1153]: okay will revert it it was to avoid flake8 errors

File: tests/docker/.env.sh (hunk 5 above)
Comment (mahajanadhitya) [reply tests/docker/.env.sh:3]: hmm , should I ask it in his office hours ?

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 8]
```diff
@@ -77,8 +81,8 @@ def __init__(self, url, max_schemas_per_subject=1000, ca_location=None, cert_loc
                 "CachedSchemaRegistry constructor is being deprecated. "
                 "Use CachedSchemaRegistryClient(dict: config) instead. "
                 "Existing params ca_location, cert_location and key_location will be replaced with their "
-                "librdkafka equivalents as keys in the conf dict: `ssl.ca.location`, `ssl.certificate.location` and "
-                "`ssl.key.location` respectively",
+                "librdkafka equivalents as keys in the conf dict: `ssl.ca.location`, `ssl.certificate.location`,  "
+                "`ssl.key.location` and `ssl.key.password` respectively",
```
[the remark this answers is not in the document]
Comment (mahajanadhitya) [reply src/confluent_kafka/avro/cached_schema_registry_client.py:85]: Did not get you on this one ?

File: src/confluent_kafka/avro/cached_schema_registry_client.py (hunk 2 above)
Comment (mahajanadhitya) [reply src/confluent_kafka/avro/cached_schema_registry_client.py:107]: This cannot be empty if you want a SSL connection, not even in the previous requests module.

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 9]
```diff
@@ -100,14 +104,22 @@ def __init__(self, url, max_schemas_per_subject=1000, ca_location=None, cert_loc
         # subj => { schema => version }
         self.subject_to_schema_versions = defaultdict(dict)
 
+        ca_certs = conf.pop('ssl.ca.location', None)
         s = Session()
```
[the remark this answers is not in the document]
Comment (mahajanadhitya) [reply src/confluent_kafka/avro/cached_schema_registry_client.py:108]: yes you are correct but would not that break any kind of backward compatibility.

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 10]
```diff
@@ -52,7 +54,9 @@ class CachedSchemaRegistryClient(object):
     .. deprecated:: 1.1.0
 
     Use CachedSchemaRegistryClient(dict: config) instead.
-    Existing params ca_location, cert_location and key_location will be replaced with their librdkafka equivalents:
+    The Config(dict) will have url, ssl.ca.location, ssl.certificate.location, ssl.key.location and ssl.key.password
```
[the remark this answers is not in the document]
Comment (mahajanadhitya) [reply src/confluent_kafka/avro/cached_schema_registry_client.py:57]: Actually it is jumbled  via the git diff , is as supposed your suggestion only. Take a look down .

**Review by pranavrth — CHANGES_REQUESTED** [review]

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 11]
```diff
@@ -158,6 +190,14 @@ def _send_request(self, url, method='GET', body=None, headers={}):
         if method not in VALID_METHODS:
             raise ClientError("Method {} is invalid; valid methods include {}".format(method, VALID_METHODS))
 
+        if url.startswith('https') and self._https_session.protected:
```
[the remark this answers is not in the document]
Comment (mahajanadhitya) [reply src/confluent_kafka/avro/cached_schema_registry_client.py:193]: password protected private key

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 12]
```diff
@@ -128,6 +135,31 @@ def close(self):
         # Constructor exceptions may occur prior to _session being set.
         if hasattr(self, '_session'):
             self._session.close()
+        if hasattr(self, '_https_session'):
+            self._https_session.clear()
+
+    @staticmethod
+    def _make_https_session(ca_certs, certs, auth, key_password):
+        _https_session = urllib3.PoolManager(cert_reqs='CERT_REQUIRED', ca_certs=ca_certs, cert_file=certs[0],
```
[the remark this answers is not in the document]
Comment (mahajanadhitya) [reply src/confluent_kafka/avro/cached_schema_registry_client.py:143]: Done

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 13]
```diff
@@ -128,6 +135,31 @@ def close(self):
...
+                                             key_file=certs[1], key_password=key_password)
+        _https_session.auth = auth
+        _https_session.protected = False
+        if key_password is not None:
+            _https_session.protected = True
+        return _https_session
+
+    @staticmethod
+    def _make_attributes_urllib3(headers, body, auth):
```
[the remark this answers is not in the document]
Comment (mahajanadhitya) [reply src/confluent_kafka/avro/cached_schema_registry_client.py]: Done!

File: src/confluent_kafka/avro/cached_schema_registry_client.py (hunk 1 above)
Comment (mahajanadhitya) [reply src/confluent_kafka/avro/cached_schema_registry_client.py:198]: Done !

File: src/confluent_kafka/avro/cached_schema_registry_client.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (pranavrth) [reply src/confluent_kafka/avro/cached_schema_registry_client.py:143]: You removed it from the wrong places.

[the remark this answers is not in the document]
Comment (pranavrth) [reply src/confluent_kafka/avro/cached_schema_registry_client.py:143]: All the private functions and fields of a class should have _. No need in the local variables.

File: tests/docker/.env.sh [hunk 14]
```diff
@@ -8,8 +8,10 @@ export PY_DOCKER_CONF=$PY_DOCKER_SOURCE/conf
 export TLS=$PY_DOCKER_CONF/tls
 
 export MY_BOOTSTRAP_SERVER_ENV=localhost:29092
-export MY_SCHEMA_REGISTRY_URL_ENV=http://$(hostname):8081
-export MY_SCHEMA_REGISTRY_SSL_URL_ENV=https://$(hostname -f):8082
+export MY_SCHEMA_REGISTRY_URL_ENV=http://localhost:8081
```
Comment (pranavrth) [inline tests/docker/.env.sh:11]: Revert the changes related to localhost in this file as well.

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 15]
```diff
@@ -128,6 +135,31 @@ def close(self):
...
+        _https_session = urllib3.PoolManager(cert_reqs='CERT_REQUIRED', ca_certs=ca_certs, cert_file=certs[0],
+                                             key_file=certs[1], key_password=key_password)
+        _https_session.auth = auth
+        _https_session.protected = False
+        if key_password is not None:
+            _https_session.protected = True
+        return _https_session
+
+    @staticmethod
+    def make_attributes_https_session(headers, body, auth):
```
Comment (pranavrth) [inline src/confluent_kafka/avro/cached_schema_registry_client.py:152]: meaningful name.

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 16]
```diff
@@ -21,6 +21,8 @@
 #
 import logging
 import warnings
+import urllib3
```
Comment (pranavrth) [inline src/confluent_kafka/avro/cached_schema_registry_client.py:24]: For now, its fine to use both urllib3 and requests. Create a backlog to move everything from requests to urllib3.

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 17]
```diff
@@ -109,6 +112,10 @@ def __init__(self, url, max_schemas_per_subject=1000, ca_location=None, cert_loc
         self.url = utils.urldefragauth(self.url)
 
         self._session = s
+        key_password = conf.pop('ssl.key.password', None)
+        _https_session = self.make_https_session(ca_path, s.cert, s.auth, key_password)
```
Comment (pranavrth) [inline src/confluent_kafka/avro/cached_schema_registry_client.py:116]: no need for this local variable. Can directly assign to `self._https_session`

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 18]
```diff
@@ -128,6 +135,31 @@ def close(self):
         # Constructor exceptions may occur prior to _session being set.
         if hasattr(self, '_session'):
             self._session.close()
+        if hasattr(self, '_https_session'):
+            self._https_session.clear()
+
+    @staticmethod
+    def make_https_session(ca_certs, certs, auth, key_password):
```
Comment (pranavrth) [inline src/confluent_kafka/avro/cached_schema_registry_client.py:142]: `ca_certs_path`

File: tests/integration/integration_test.py [hunk 19]
```diff
@@ -1254,6 +1254,16 @@ def print_usage(exitcode, reason=None):
     if 'avro-https' in modes:
         print('=' * 30, 'Verifying AVRO with HTTPS', '=' * 30)
         verify_avro_https(testconf.get('avro-https', None))
+        conf = testconf.get("avro-https-key-with-password", None)
```
Comment (pranavrth) [inline tests/integration/integration_test.py:1257]: 2 confs are being used in this function. use `key-with-password-conf`.

File: src/confluent_kafka/avro/cached_schema_registry_client.py (hunk 18 above)
Comment (pranavrth) [reply src/confluent_kafka/avro/cached_schema_registry_client.py:142]: Use two different parameters for certificate and key. Don't use `certs` to represent both.

File: src/confluent_kafka/avro/cached_schema_registry_client.py (hunk 17 above)
Comment (mahajanadhitya) [reply src/confluent_kafka/avro/cached_schema_registry_client.py:116]: Sure!

File: src/confluent_kafka/avro/cached_schema_registry_client.py (hunk 16 above)
Comment (mahajanadhitya) [reply src/confluent_kafka/avro/cached_schema_registry_client.py:24]: Sure

File: src/confluent_kafka/avro/cached_schema_registry_client.py (hunk 18 above)
Comment (mahajanadhitya) [reply src/confluent_kafka/avro/cached_schema_registry_client.py:142]: Done !

File: tests/integration/integration_test.py (hunk 19 above)
Comment (mahajanadhitya) [reply tests/integration/integration_test.py:1257]: just change the name of conf right ?

Comment (mahajanadhitya) [reply tests/integration/integration_test.py:1257]: Thats Done!

File: src/confluent_kafka/avro/cached_schema_registry_client.py (hunk 15 above)
Comment (mahajanadhitya) [reply src/confluent_kafka/avro/cached_schema_registry_client.py:152]: Working on it!

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 20]
```diff
@@ -128,6 +136,29 @@ def close(self):
         # Constructor exceptions may occur prior to _session being set.
         if hasattr(self, '_session'):
             self._session.close()
+        if hasattr(self, '_https_session'):
+            self._https_session.clear()
+
+    @staticmethod
+    def make_https_session(ca_certs_path, cert_location, key_location, auth, key_password):
+        _https_session = urllib3.PoolManager(cert_reqs='CERT_REQUIRED', ca_certs=ca_certs_path,
```
Comment (pranavrth) [inline src/confluent_kafka/avro/cached_schema_registry_client.py:144]: Remove _

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 21]
```diff
@@ -128,6 +136,29 @@ def close(self):
         # Constructor exceptions may occur prior to _session being set.
         if hasattr(self, '_session'):
             self._session.close()
+        if hasattr(self, '_https_session'):
+            self._https_session.clear()
+
+    @staticmethod
+    def make_https_session(ca_certs_path, cert_location, key_location, auth, key_password):
```
Comment (pranavrth) [inline src/confluent_kafka/avro/cached_schema_registry_client.py:143]: Add _

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 22]
```diff
@@ -128,6 +136,29 @@ def close(self):
...
+        _https_session = urllib3.PoolManager(cert_reqs='CERT_REQUIRED', ca_certs=ca_certs_path,
+                                             cert_file=cert_location, key_file=key_location, key_password=key_password)
+        _https_session.auth = auth
+        return _https_session
+
+    def send_https_session_request(self, url, method, headers, body):
```
Comment (pranavrth) [inline src/confluent_kafka/avro/cached_schema_registry_client.py:149]: Add _

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 23]
```diff
@@ -128,6 +136,29 @@ def close(self):
...
+        _headers = {'Accept': ACCEPT_HDR}
```
Comment (pranavrth) [inline src/confluent_kafka/avro/cached_schema_registry_client.py:150]: Remove _

File: src/confluent_kafka/avro/cached_schema_registry_client.py (hunk 21 above)
Comment (pranavrth) [reply src/confluent_kafka/avro/cached_schema_registry_client.py:143]: change param ordering

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 24]
```diff
@@ -109,6 +112,11 @@ def __init__(self, url, max_schemas_per_subject=1000, ca_location=None, cert_loc
         self.url = utils.urldefragauth(self.url)
 
         self._session = s
+        key_password = conf.pop('ssl.key.password', None)
+        self.passwordprotected = False
```
Comment (pranavrth) [inline src/confluent_kafka/avro/cached_schema_registry_client.py:116]: is_key_password_provided

**Review by pranavrth — CHANGES_REQUESTED** [review]

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 25]
```diff
@@ -109,6 +112,11 @@ def __init__(self, url, max_schemas_per_subject=1000, ca_location=None, cert_loc
         self.url = utils.urldefragauth(self.url)
 
         self._session = s
+        key_password = conf.pop('ssl.key.password', None)
+        self.is_key_password_provided = False
```
Comment (pranavrth) [inline src/confluent_kafka/avro/cached_schema_registry_client.py:116]: 1. use _.
2. Also do an empty string check.

Assignment of `_is_key_password_provided` can be converted to `self._is_key_password_provided = not key_password` expression.

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 26]
```diff
@@ -128,6 +136,29 @@ def close(self):
...
+        https_session.auth = auth
+        return https_session
+
+    def _send_https_session_request(self, url, method, headers, body):
+        request_headers = {'Accept': ACCEPT_HDR}
+        auth = self._https_session.auth
+        if body:
+            body = json.dumps(body).encode('UTF-8')
+            request_headers["Content-Length"] = str(len(body))
+            request_headers["Content-Type"] = "application/vnd.schemaregistry.v1+json"
```
Comment (pranavrth) [inline src/confluent_kafka/avro/cached_schema_registry_client.py:155]: This should be application/json as body is being converted to Json in the above call. Please check.

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 27]
```diff
@@ -158,6 +189,13 @@ def _send_request(self, url, method='GET', body=None, headers={}):
         if method not in VALID_METHODS:
             raise ClientError("Method {} is invalid; valid methods include {}".format(method, VALID_METHODS))
 
+        if url.startswith('https') and self.passwordprotected:
+            response = self.send_https_session_request(url, method, headers, body)
```
Comment (pranavrth) [inline src/confluent_kafka/avro/cached_schema_registry_client.py:193]: _send_https_session_request

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 28]
```diff
@@ -158,6 +189,13 @@ def _send_request(self, url, method='GET', body=None, headers={}):
         if method not in VALID_METHODS:
             raise ClientError("Method {} is invalid; valid methods include {}".format(method, VALID_METHODS))
 
+        if url.startswith('https') and self.passwordprotected:
```
Comment (pranavrth) [inline src/confluent_kafka/avro/cached_schema_registry_client.py:192]: There is nothing like `passwordprotected` anymore? Please configure your IDE to identify these kind of problems.

**Review by pranavrth — CHANGES_REQUESTED** [review]

**Comment by pranavrth** [comment]
Fix build

File: src/confluent_kafka/avro/cached_schema_registry_client.py (hunk 26 above)
Comment (mahajanadhitya) [reply src/confluent_kafka/avro/cached_schema_registry_client.py:155]: Actually it has to do with the interal working of the schema registry and was this for all the request with the body for the old flow as well!

File: src/confluent_kafka/avro/cached_schema_registry_client.py (hunk 25 above)
Comment (mahajanadhitya) [reply src/confluent_kafka/avro/cached_schema_registry_client.py]: sure

File: src/confluent_kafka/avro/cached_schema_registry_client.py [hunk 29]
```diff
@@ -109,6 +112,9 @@ def __init__(self, url, max_schemas_per_subject=1000, ca_location=None, cert_loc
         self.url = utils.urldefragauth(self.url)
 
         self._session = s
+        key_password = conf.pop('ssl.key.password', None)
+        self.is_key_password_provided = not key_password
```
Comment (pranavrth) [inline src/confluent_kafka/avro/cached_schema_registry_client.py:116]: Add '_'. Change `self.is_key_password_provided` to `self._is_key_password_provided` as this is going to be a private variable.

**Review by pranavrth — APPROVED** [review]
LGTM!
